### ISSUE TITLE
`use_stateful_dataloader` + `num_workers>0` draws epoch-0 permutation during `prepare()` before cross-rank RNG sync — corrupts DDP data partition

### ISSUE THREAD

**Issue by GoldenStain** [huggingface/accelerate#4204]
### System Info

```Shell
- `accelerate` version: 1.13.0 (and main branch)
- `torch` version: 2.13.0
- `torchdata` version: 0.11.0
- Platform: Linux
- Python version: 3.11
- Distributed environment: 2-process CPU/Gloo (reproducible without GPU)
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [x] My own task or dataset (give details below)

#### Summary & Root Cause

With `DataLoaderConfiguration(use_stateful_dataloader=True)`, `accelerator.prepare()` wraps the dataloader as a torchdata `StatefulDataLoader` and immediately invokes `base_dataloader.state_dict()` inside `DataLoaderAdapter.__init__` to record the initial dataloader state.

When `num_workers > 0`:
1. `StatefulDataLoader.state_dict()` calls `self._get_iterator()` when `self._iterator is None` to snapshot iterator state.
2. The multiprocessing iterator's `_reset()` spawns worker processes and primes the prefetch queue (`prefetch_factor * num_workers` items), which iterates `_sampler_iter`.
3. For a vanilla `RandomSampler(generator=None)`, this immediately draws the epoch-0 permutation seed from the process's **local global RNG at `prepare()` time**.
4. Cross-rank dataloader RNG synchronization (`DataLoaderShard.__iter__` -> `synchronize_rng_states`) only happens when iteration actually starts, so any divergence in per-rank RNG before `prepare()` (e.g. no global `torch.manual_seed` set, deliberate per-rank seeding, or unequal RNG consumption across ranks during model/dataset setup) causes ranks to derive **different** epoch-0 permutations.
5. `BatchSamplerShard` then rank-interleaves divergent permutations, which **corrupts the data partition during epoch 0** (samples are duplicated across ranks and others are dropped).

*(Note: This affects uninterrupted training runs during epoch 0, independent of any checkpoint/resume operation).*

#### Minimal Reproducible Example

Run with 2 processes on CPU (no GPU needed):

```bash
# Case A (diverged/unseeded RNG across ranks): partition is corrupted
python -m torch.distributed.run --standalone --nproc_per_node=2 repro_prepare_prefetch.py noseed

# Case B (identical seed explicitly set before prepare): partition is intact
python -m torch.distributed.run --standalone --nproc_per_node=2 repro_prepare_prefetch.py seed
```

`repro_prepare_prefetch.py`:

```python
import sys
import torch
import torch.distributed as dist
from torch.utils.data import DataLoader, TensorDataset
from accelerate import Accelerator, DataLoaderConfiguration

MODE = sys.argv[1]  # "seed" or "noseed"
N, BS = 96, 4

if MODE == "seed":
    torch.manual_seed(1234)  # identical on both ranks before prepare()

acc = Accelerator(cpu=True, dataloader_config=DataLoaderConfiguration(use_stateful_dataloader=True))
loader = acc.prepare(
    DataLoader(TensorDataset(torch.arange(N)), batch_size=BS, shuffle=True, num_workers=6)
)

# Iterate through epoch 0
flat = [i for b in loader for i in b[0].tolist()]

gathered = [None] * acc.num_processes
dist.all_gather_object(gathered, flat)
acc.wait_for_everyone()

if acc.is_main_process:
    union = [i for rank_items in gathered for i in rank_items]
    union_set = set(union)
    missing = len(set(range(N)) - union_set)
    duplicates = len(union) - len(union_set)
    print(
        f"mode={MODE}: epoch-0 union={len(union_set)}/{N}, missing={missing}, duplicates={duplicates}"
    )
```

#### Observed Output

```
mode=noseed: epoch-0 union=72/96, missing=24, duplicates=24
mode=seed:   epoch-0 union=96/96, missing=0, duplicates=0
```

With `num_workers=6` and no prior shared seed synchronization, 24 out of 96 samples are duplicated across ranks and 24 samples are completely omitted from epoch 0.

### Expected behavior

`accelerator.prepare(dataloader)` should not draw from the sampler or consume generator RNG before iteration begins:

1. **Lazy state capture (Preferred)**: `DataLoaderAdapter.__init__` should avoid materializing `_get_iterator()` during initialization, capturing state lazily on the first step/checkpoint instead.
2. **Post-init cleanup / reset**: If `state_dict()` must be called during `prepare_data_loader`, the early iterator should be cleared (`dataloader.base_dataloader._iterator = None` or equivalent) to reap worker pools and defer the sampler draw to iteration time.
3. At minimum, `synchronize_rng_states` should run prior to adapter initial state capture if `use_stateful_dataloader=True` and `num_workers > 0`.

**Comment by jayzuccarelli** [comment]
Confirmed on `main`: `DataLoaderAdapter.__init__` captures `self.dl_state_dict = self.base_dataloader.state_dict()` eagerly (`src/accelerate/data_loader.py`, right after the `StatefulDataLoader` is built). With `num_workers > 0`, `StatefulDataLoader.state_dict()` has to build an iterator to snapshot it, and building the iterator spawns the workers and draws the epoch-0 permutation from the local RNG. `synchronize_rng_states` only runs later, at the top of `DataLoaderShard.__iter__`, so the permutation is already fixed per rank by then.

Smallest fix I see: capture lazily. Set `self.dl_state_dict = None` in `__init__`, and have `state_dict()` fall back to `self.base_dataloader.state_dict()` when nothing has been captured yet (a checkpoint taken before the first batch keeps working, and `_update_state_dict()` overwrites it as soon as iteration starts). That moves the first iterator creation into `__iter__`, after the RNG sync, and doesn't touch the resume path (#4195).

The alternative is to keep the eager capture and call `synchronize_rng_states` before it, but `DataLoaderAdapter` doesn't know `rng_types` and the sync would then run twice per epoch 0.

@SunMarc happy to open a PR for the lazy version with a 2-process test based on the repro above, if that's the direction you want.

**Comment by SunMarc** [comment]
Thanks for the report ! I'll try to investigate myself. I closed the PRs above as they are slop code agents. I encourage you to try fix the issue yourself if you have time also, happy to review @GoldenStain

**Comment by jayzuccarelli** [comment]
@SunMarc happy to take this if @GoldenStain isn't already on it. Lazy capture in `DataLoaderAdapter` + a 2-process test for the rank divergence, can have a PR up in a day.

**Comment by SeverinVisionary** [comment]
One more consequence of the eager `state_dict()` in `DataLoaderAdapter.__init__`, relevant to #4234. It also breaks resume, even when the seed is identical on every rank.

On resume, `prepare()` has already built an iterator and drawn a permutation. `load_state_dict()` then builds a second iterator, so the resumed process draws one permutation too many and continues from the saved position in the wrong order. This happens in epoch 0 as well.

```python
# python repro.py save && python repro.py resume   (or: torchrun --nproc_per_node=2 repro.py save|resume)
import sys

import torch
from torch.utils.data import DataLoader

from accelerate import Accelerator, DataLoaderConfiguration


def main():
    mode, num_workers = sys.argv[1], 2
    torch.manual_seed(1234)  # identical on every rank
    acc = Accelerator(cpu=True, dataloader_config=DataLoaderConfiguration(use_stateful_dataloader=True))
    dl = acc.prepare(DataLoader(list(range(96)), batch_size=4, shuffle=True, num_workers=num_workers))
    rank = acc.process_index
    if mode == "save":
        it = iter(dl)
        for _ in range(5):
            next(it)
        torch.save(dl.state_dict(), f"state{rank}.pt")
        torch.save([b.tolist() for b in it], f"rest{rank}.pt")  # the true remainder of epoch 0
    else:
        dl.load_state_dict(torch.load(f"state{rank}.pt", weights_only=False))
        got, rest = [b.tolist() for b in dl], torch.load(f"rest{rank}.pt")
        print(f"rank {rank}: resumed remainder == true remainder: {got == rest}")


if __name__ == "__main__":
    main()
```

On accelerate 1.15.0, torch 2.14.0, torchdata 0.11.0 (CPU), this prints `False` for one process and for both ranks of a 2-process run.

Resuming from 5 batches into epoch 0 and comparing with the true remainder:

| Loader | 1 process, 0 workers | 1 process, 2 workers | 2 processes, 0 workers | 2 processes, 2 workers |
|---|---|---|---|---|
| torchdata `StatefulDataLoader` alone | exact | exact | exact | exact |
| accelerate-prepared | differs | differs (next epoch's permutation at the cursor) | exact | differs |
| accelerate-prepared, `StatefulDataLoader.state_dict` stubbed out during `prepare()` (i.e. #4234's lazy capture) | exact | exact | exact | exact |

So the lazy capture in #4234 fixes the resume case too, including single-process with `num_workers=0`. A regression test that saves partway through epoch 0 and resumes in a fresh process with `num_workers > 0` would cover it.
